### PR TITLE
Smarter push/pull fallback from TLS to plaintext

### PR DESCRIPTION
**Pull request by aaronlehmann**
With the --insecure-registry daemon option (or talking to a registry on a local IP), the daemon will first try TLS, and then try plaintext if something goes wrong with the push or pull. It doesn't make sense to try plaintext if a HTTP request went through while using TLS. This commit changes the logic to keep track of host/port combinations where a TLS attempt managed to do at least one HTTP request (whether the response code indicated success or not). If the host/port responded to a HTTP using TLS, we won't try to make plaintext HTTP requests to it.

This will result in better error messages, which sometimes ended up showing the result of the plaintext attempt, like this:

```
Error response from daemon: Get
http://myregistrydomain.com:5000/v2/: malformed HTTP response
"\x15\x03\x01\x00\x02\x02"
```

This fixes one of the things reported in https://github.com/docker/docker/issues/19834.

cc @docker/distribution-maintainers

### CODE REVIEW

**Comment by vikstrous** [comment]
Sweet. I just tried this. It seems to work as advertised. It also fixes https://github.com/docker/docker/issues/20104 I think

**Comment by cpuguy83** [comment]
Just setting to design review since it's not ready and we don't have a better label right now.

**Comment by vikstrous** [comment]
One more awkward thing is left. It's still not obvious what's going on if you try to push to a server that isn't listening on any ports. It looks like the connection on port 80 faild, but really the issue is that port 443 failed. Can we preserve the error message from the first failure?

```
$ docker push loc2.loc/admin/repo
The push refers to a repository [loc2.loc/admin/repo]
Put http://loc2.loc/v1/repositories/admin/repo/: dial tcp 192.168.1.2:80: getsockopt: connection refused
```

**Comment by tiborvass** [comment]
Agree with https://github.com/docker/docker/pull/20253#issuecomment-183153277 otherwise this is a really welcome improvement.

**Comment by runcom** [comment]
apart from https://github.com/docker/docker/pull/20253#issuecomment-183153277 this is fixing my issue :)

**Comment by aaronlehmann** [comment]
Rebased; removed WIP note.

The bit about servers that aren't listening on either port is interesting. For much of the 1.10 dev cycle, we displayed the errors from all the attempts. This was found to be confusing in some cases and the change was reverted before the 1.10 release. See https://github.com/docker/docker/pull/19424

I'd like to find a solution to this, but I think it's better for this to be a separate PR, since it's more about error reporting than controlling the actual connection attempts. Could you please file an issue for tracking purposes?

File: distribution/push_v2.go [hunk 1]
```diff
@@ -61,15 +61,24 @@ type pushState struct {
...
+		}
 	}
 
 	if err = p.pushV2Repository(ctx); err != nil {
 		if continueOnError(err) {
-			return fallbackError{err: err, confirmedV2: p.pushState.confirmedV2}
+			return fallbackError{
+				err:         err,
+				confirmedV2: p.pushState.confirmedV2,
+				transportOK: true,
```
Comment (stevvooe) [inline distribution/push_v2.go]: Should this use the variable value from above?

File: distribution/registry.go [hunk 2]
```diff
@@ -60,14 +60,17 @@ func NewV2Repository(ctx context.Context, repoInfo *registry.RepositoryInfo, end
 	endpointStr := strings.TrimRight(endpoint.URL, "/") + "/v2/"
 	req, err := http.NewRequest("GET", endpointStr, nil)
 	if err != nil {
-		return nil, false, err
+		return nil, false, false, err
```
Comment (stevvooe) [inline distribution/registry.go]: Is there anyway we can embed this in the error type? I am reticent in expanding the number of return values from this function.

Let me know if you need ideas.

Comment (aaronlehmann) [inline distribution/registry.go]: @stevvooe: I've updated the PR to use a different approach, please tell me if you think it's an improvement. Now `NewV2Repository` returns a `fallbackError` directly instead of depending on the caller to wrap the error it returns. This means we can skip the extra return value.

File: distribution/push_v2.go (hunk 1 above)
Comment (aaronlehmann) [inline distribution/push_v2.go]: > Should this use the variable value from above?

In the original version, `NewV2Repository` always returned `transportOK == true` on success, because completing this function successfully involves a successful HTTP "ping". Per your other comment, I've updated it to return `transportOK` as part of the error it returns. This should make it clearer that `transportOK` is only false in certain cases where `NewV2Repository` fails.

Comment (stevvooe) [inline distribution/push_v2.go]: Much cleaner.

**Comment by stevvooe** [comment]
LGTM

**Comment by dmcgowan** [comment]
LGTM

**Comment by thaJeztah** [comment]
moving to code review (@stevvooe @dmcgowan pls re-LGTM of code is ok; not sure if your LGTM's were on design or code)

File: distribution/push.go [hunk 3]
```diff
@@ -127,6 +133,19 @@ func Push(ctx context.Context, ref reference.Named, imagePushConfig *ImagePushCo
 			continue
 		}
 
+		parsedURL, urlErr := url.Parse(endpoint.URL)
```
Comment (calavera) [inline distribution/push.go]: can we extract this logic to a function? it looks like it's duplicated in `pull.go`.

Comment (aaronlehmann) [inline distribution/push.go]: @calavera: This is a fair point - quite possibly one I would raise myself if I reviewed similar code. I thought about this while working on the PR, but didn't immediately see a clean way to split the code out.

The new code here involves parsing the URL and also maintaining a map. The direct way to reuse it would be to provide a function that parses the URL, checks for it in the map, and returns the parsed representation, and also a function that adds the URL to the map if appropriate. Both functions would have to take a fair amount of state in arguments or a special structure, and it seemed like it would be ugly.

The `Push` and `Pull` functions are nearly identical, except one calls `newPuller`/`Pull` and the other calls `newPusher`/`Push` (and they have log messages tailored to the particular operation). Ideally `Push` and `Pull` would be combined into something more generic. The obstacle is that `Pusher` and `Puller` are different interfaces, so I don't see an obvious way to write a clean generic version.

I will experiment with this and get back to you.

Comment (stevvooe) [inline distribution/push.go]: @calavera Do you mean something like `RequiresTLS(url) (bool, error)`?

Comment (aaronlehmann) [inline distribution/push.go]: @calavera: Do you prefer this approach? https://gist.github.com/aaronlehmann/dcec73055470ba81cd4f

I'm not sure I like the set of function arguments and return values:

    func checkRegistryIncompatibility(endpoint registry.APIEndpoint, confirmedTLSRegistries map[string]struct{}, confirmedV2 bool) (*url.URL, error) {

...but I don't really have strong feelings about this.

Comment (stevvooe) [inline distribution/push.go]: I don't really think that adds clarity. The current code parses a url and checks the scheme. This is very clear. No need to be DRY for DRY's sake.

Comment (calavera) [inline distribution/push.go]: @aaronlehmann 

https://play.golang.org/p/IBM0WZkHV1

Comment (aaronlehmann) [inline distribution/push.go]: @calavera: I guess there would also be a method to mark a registry as a confirmed TLS registry? Where should the URL parsing happen for this? Should both methods have their own redundant calls to `url.Parse`?

Comment (calavera) [inline distribution/push.go]: > Where should the URL parsing happen for this?

I really don't know why `APIEndpoint` stores `URL` as `string` and `Endpoint` stores `URL` as `*url.URL` but that looks like a smell tbh. Why don't we have a single place where all the parsing happens? Would that make all this logic easier?

Comment (aaronlehmann) [inline distribution/push.go]: @calavera: I agree that it would make more sense for `APIEndpoint` to contain a parsed URL. I added a commit that makes this change.

Please let me know whether:

a) the change looks good
b) this should be broken out into a separate PR

**Comment by tiborvass** [comment]
LGTM, the win2lin issue is known and being worked on.
